rotary/repair.py

```diff
@@ -15,14 +15,15 @@
 import pandas as pd
 
 # GLOBAL VARIABLES
-SCRIPT_VERSION = '0.2.0'
 DEPENDENCY_NAMES = ['flye', 'minimap2', 'samtools', 'circlator']
 
 # Set up the logger
-logging.basicConfig(format='[ %(asctime)s UTC ]: %(module)s: %(funcName)s: %(levelname)s: %(message)s')
-logging.Formatter.converter = time.gmtime
 logger = logging.getLogger(__name__)
-
+logger.setLevel(logging.DEBUG)
```

Comment:
@jmtsuji Okay you already do this with the ```--verbose``` flag. Great job! I would still set the default flag to info rather than debug. P.s. As discussed in other comments, I would have seen this flag earlier if ```main()``` was at the top of the file.

Reply:
@LeeBergstrand Interestingly, I found that the default logger level needs to be set to the most verbose level that you want to be able to set the log handlers to. See the flow chart here: https://docs.python.org/3/howto/logging.html#logging-flow

The repair code has two log handlers: `StreamHandler()` that writes to screen and `FileHandler()` that writes to a file. As shown in the flow chart, if you set the "root" logger level to INFO, it will drop anything at DEBUG level before it has a chance to get to either of the handlers. I tested this, and indeed, you can't log at DEBUG (even if you set the handlers to DEBUG level) if the "root" logger level is INFO.

With the current setup in the repair code, the user can control the verbosity level that gets printed to the screen (via `StreamHandler()`) using the `--verbose` flag. I forced the log file (via `FileHandler()`) to always write at DEBUG in case the user wanted to check on details of the run at some point in the future. Because of this, it made sense to hard-code the "root" logger level to DEBUG outside of main().

Do you think it would be better to let the user control the verbosity of both the screen output and the file log via the `--verbose` flag? If so, then it might be possible to move the "root" logger level setting into main() so that it doesn't need to be hard-coded as DEBUG. I hope this somewhat convoluted explanation makes sense. Thanks!

Reply:
@jmtsuji Thanks for the info regarding setting the initial log level to debug at the top and then reducing it later. Great to know.

**Context**: So at the production scale, you would have something like Rotary run in response to a cron job or a push notification. So, for example, when a new genome gets uploaded to the data server or Google Cloud, it would automatically run Rotary from a script on that genome, or you would rerun the entire catalogue of genomes every six months with the latest databases. This is an approach that MLBS might be taking down the road and something we are already doing with 16S seq. If a program is left in debug mode, you would be surprised how much space log files take and how long it takes to analyze them. At scale, the companies start sending logs back to a centralized server where they can be analyzed: https://www.datadoghq.com/blog/python-logging-best-practices/. You can monitor these centralized logs to get performance information from the time stamps (how long tasks take in deployment). In addition, in production, most log files are set up to roll over at a fixed amount of space or time, so they will only hold the last few days of logging. So, a lowered level of logging can lead to more days of logging stored for analysis. The standard operating procedure is leaving both the visual and written logs non-verbose, where they get logged back to files or a central server. If something goes wrong, you can go through the logs to find the failed data, and then the investigation team can download the failed data and rerun the pipeline in verbose mode to figure out what went wrong and apply fixes. The log messages should have the file names or other identifying parameters used to determine which files failed to support such investigations.

> Do you think it would be better to let the user control the verbosity of both the screen output and the file log via the `--verbose` flag? If so, then it might be possible to move the "root" logger level setting into main() so that it doesn't need to be hard-coded as DEBUG. I hope this somewhat convoluted explanation makes sense. Thanks!

Yes, I think using the ```--verbose``` flag should control the log level of both the screen output and the log file.

Reply:
This is really helpful additional context -- thanks! I have updated the code as discussed.